front: import osrd-ui

emersion opened the pull request:
_See individual commits._

TODO:

- [x] Basic NPM workspace setup
- [x] Build system
- [x] Docker integration
  - [x] Base flavor
  - [x] Front flavor
- [x] Storybook publishing
- [x] Releases
- [x] Adjust docs

Closes: https://github.com/OpenRailAssociation/osrd-ui/issues/451

## Design decisions

### Why move `front/ui/package.json` to `front/ui/base/package.json`?

NPM doesn't correctly handle nested sub-packages (`front/ui/` and `front/ui/storybook/`, for instance). It doesn't install dev dependencies of nested sub-packages (storybook dependencies are missing).

### Why keep `front/ui/base/package.json`?

`front/ui/` contains config files for rollup, ESLint, postcss used by all ui-* sub-packages. We need to specify dependencies for these somewhere.

We could specify them in `front/package.json`, but that would mix up osrd and osrd-ui deps. Keeping shared osrd-ui deps inside `front/ui/base/package.json` makes it clearer what's required by osrd itself and what's required by osrd-ui.

### Should we build osrd-ui when running `npm run build`?

Another possible solution would've been to only build osrd when running `npm run build`, and require users to run `npm run build --workspaces` to build osrd-ui. The upside would be that less work is performed when only touching osrd.

However, this would make it so `npm run build` would fail if osrd-ui hasn't been built before, or has an outdated build. Also, building osrd-ui requires less changes to the CI and tooling.

### Should we build osrd-ui's storybook when building osrd's frontend?

The osrd-ui storybook is built when running `npm run build` in `front/`, but is not required for the frontend.

This simplifies a bit CI, because there is a single build/test/lint command to execute instead of separate commands for osrd itself and the storybook. If we choose to only execute the bare minimum there might be some inconsistencies between the scripts: build will need to build osrd-ui (without the storybook) but test/lint would not operate on osrd-ui. Additionally, `npm run build` in `front/` is seldom used outside of CI, and the current approach requires less changes in the CI manifests.

One could argue that we should not build the storybook when running `npm run build`. There are already some discrepancies in the scripts, e.g. `compile` and `lint-fix` don't operate on osrd-ui, and `start` requires osrd-ui to be already built.

### Why not build osrd-ui on startup for the front flavor of the Compose setup?

Right now, the container might spew some errors the first time it's started because it's missing an osrd-ui build. The osrd-ui build happens in parallel, so the errors should eventually go away, except for ESLint which unfortunately cannot watch osrd-ui. Restarting the container should fix the ESLint issues. This should only happen the very first time the front container is started.

We could build osrd-ui before starting osrd, but this would add 30s or 1min of extra delay every time the container starts up. This was deemed too annoying, especially if restarting the container fixes it.

### There are ESLint errors when I start the front flavor of the Compose setup

This is due to osrd-ui and osrd being built in parallel when the container starts, and ESLint doesn't notice when osrd-ui has been built. Indeed, ESLint doesn't install a watcher on files in `node_modules/`, which osrd-ui deps point to.

To fix, run `npm run build --workspaces` before starting the container, or restart the container (either should work).

### How will we merge this?

We will need to coordinate a bit to not loose anything in osrd-ui. Easiest would be to get this reviewed and ready to merge, freeze osrd-ui, release a new osrd-ui version and bump osrd to use that new version (if there are breaking API changes), adjust the "import osrd-ui" commit to import that version, and merge this PR.

Once that's done, all opened osrd-ui PRs will need to be manually migrated to the osrd repository.

<a name="migrate-pr"></a>
One can migrate a PR from osrd-ui to osrd like so:

- Create a new branch in osrd
- Run the command below to apply the commits locally
- Push to GitHub
- Open a new PR with a link to the old osrd-ui PR in the description
- Assign reviewers

```sh
pr=976
curl -L "https://github.com/OpenRailAssociation/osrd-ui/pull/$pr.patch" | git am --directory=front/ui
```

### Why `front/ui/` instead of `front/osrd-ui/`?

Because "osrd" is already in the name of this repo  ¯\\\_(ツ)\_/¯

emersion committed:
front: add ui/ to ESLint, Prettier and TypeScript ignores

Otherwise these will blow up with out-of-memory errors when a new
ui/ directory is imported in the next commit.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: import osrd-ui

This is the osrd-ui tree imported from tag 0.0.74, without any
change:

    git clone --branch 0.0.74 https://github.com/OpenRailAssociation/osrd-ui.git ui
    rm -rf ui/.git

Git history up to this state is available at:
https://github.com/OpenRailAssociation/osrd-ui/

The author list has been generated with:

    git log --format='%aN <%aE>' | sort | uniq

With the following .mailmap file:
https://paste.sr.ht/~emersion/6cedd39f2935f6fcd3854eff1ce759c037965003

Signed-off-by: Simon Ser <contact@emersion.fr>
Co-authored-by: Achraf Mohyeddine <a.mohyeddine@gmail.com>
Co-authored-by: Alexis Jacomy <alexis.jacomy@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>
Co-authored-by: Benoit Simard <contact@bsimard.com>
Co-authored-by: Chaka NGAMENI NJINEH <ngamenichaka@yahoo.fr>
Co-authored-by: Clara Ni <clara.ni@outlook.fr>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Egor Berezovskiy <egor@berezify.fr>
Co-authored-by: Élyse Viard <elysasrc@tuta.com>
Co-authored-by: Ethan Perruzza <ethan.perruzza@gmail.com>
Co-authored-by: Florian Amsallem <florian.amsallem@gmail.com>
Co-authored-by: Gaël Haméon <gael.hameon@gmail.com>
Co-authored-by: Ian Turner <turnerian2004@gmail.com>
Co-authored-by: Mathieu <mathieu.coulibaly@sncf.fr>
Co-authored-by: Math_R_ <mathieu.richard747@gmail.com>
Co-authored-by: Nicolas Wurtz <gitlab@nethenic.net>
Co-authored-by: romainvalls <romain.valls95@gmail.com>
Co-authored-by: rostom boudi <ext.rostom.boudi@reseau.sncf.fr>
Co-authored-by: SarahBellaha <sarah.bellaha@sncf.fr>
Co-authored-by: Serge Croisé <SergeCroise@users.noreply.github.com>
Co-authored-by: SharglutDev <p.filimon75@gmail.com>
Co-authored-by: theocrsb <theo_crosbie@yahoo.fr>
Co-authored-by: Theo Macron <theo.macron0315@gmail.com>
Co-authored-by: thibautsailly <thibaut@tsailly.net>
Co-authored-by: Uriel-Sautron <uriel.sautron@gmail.com>
Co-authored-by: Valentin Chanas <anisometropie@gmail.com>
Co-authored-by: Victor "multun" Collod <victor.collod@epita.fr>
Co-authored-by: Yohh <durandyohan@zaclys.net>
Co-authored-by: Younes Khoudli <younes.khoudli@epita.fr>

emersion committed:
github: import workflow to optimize ui-icons

File paths need to be slightly adjusted.

requirements.txt is moved to the root of ui-icons so that it can
more easily be run locally.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
github: import workflow to publish osrd-ui website

The build script is run for all workspaces, but not the root
workspace.

The deployment repository now needs to be explicitly specified.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: delete duplicate osrd-ui files

All of these are already covered by the osrd monorepo.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
github: merge osrd-ui dependabot configuration

Since front includes osrd-ui via NPM workspaces, dependabot will
upgrade osrd-ui dependencies. We only need to slightly adjust
group patterns to include everything missing from osrd-ui's
configuration file.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: wire up osrd-ui to osrd front build

osrd-ui's root package.json is moved to ui/base/ because nested
workspaces don't work with NPM:
https://github.com/npm/cli/issues/4774

rollup-base.config.js needs to be adjusted because node_modules
now lives one directory higher and dist directories live one
directory deeper.

The build, test and lint scripts are adjusted to execute tasks for
both osrd and osrd-ui.

The full front source code is copied over in the Dockerfile before
installing dependencies, because NPM needs to look at all workspace's
package.json files to install dependencies.

eslint-plugin-react-hooks needs to be downgraded because osrd itself
is not ready to use v5 (it causes conflicts with eslint-config-airbnb).

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
github: import workflow to publish osrd-ui releases

We need to strip the "v" prefix from Git tag names, and adjust file
paths a bit.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: build osrd-ui in Compose front dev flavor

When starting the front dev container, osrd-ui may not have been
built yet, or might be outdated. Start a build of osrd-ui
components which are used in osrd in parallel with vite.

Avoid building osrd-ui's storybook, because it takes quite a while
and it's unnecessary.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: adjust build commands in READMEs

osrd-ui needs to be built before starting the frontend.

The "storybook" script is no more, instead users can cd into the
storybook directory (just like all other sub-packages).

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: document Docker dev setup

There was no hint to indicate how one would start Docker Compose
to work on the frontend.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: disable ESLint cache

Unfortunately, ESLint's cache is pretty bare-bones and doesn't
invalidate dependent files. For example, if a warning appears in
foo.ts but is caused by a mistake in bar.ts, fixing bar.ts will not
cause ESLint to re-check foo.ts and the warning will linger.

I've seen multiple occurences of this happening. Deleting the
ESLint cache directory and restarting ESLint helps, but it pretty
cumbersome.

Slow checks are better than invalid/stale warnings, so let's switch
back to no cache.

See the typescript-eslint docs, which also recommend to disable
the cache:
https://typescript-eslint.io/troubleshooting/faqs/eslint/#can-i-use-eslints---cache-with-typescript-eslint

Signed-off-by: Simon Ser <contact@emersion.fr>

Yohh commented on .github/workflows/ui-icons-optimize.yml:
```diff
@@ -0,0 +1,39 @@
+on:
+  # This workflow will be laucnhed manually and will trigger a new commit
+  # on the branc that is specified.
```
```suggestion
  # on the branch that is specified.
```

emersion replied:
I'd prefer to not fix any unrelated issues in files moved around. This file comes from osrd-ui and is only slightly adjusted to accommodate for the new file paths.

Yohh commented on front/README.md:
```diff
@@ -2,9 +2,26 @@
+    ./osrd-compose dev-front build
+    ./osrd-compose up -d
+
+The first time the container starts up, the osrd-ui library will be missing.
+This will trigger some build errors, which should go away as soon as osrd-ui
+gets built. Restarting the front container helps getting rid of lingering
+ESLint errors.
```
not sure to understand this, will it be the same every time OSRD is fully setup?

Yohh replied:
just tried it, I have no errors, no missing library

emersion replied:
This only happens the very first time, with a freshly cloned osrd repository. Maybe you were lucky and osrd-ui finished building before ESLint kicked in. (It's all done in parallel, so either may win the race.)

Yohh commented on front/package.json:
```diff
@@ -145,20 +153,26 @@
     "i18n-api-errors": "tsx ./scripts/i18n-api-errors.ts",
-    "i18n-checker": "tsx ./scripts/i18n-checker.ts"
+    "i18n-checker": "tsx ./scripts/i18n-checker.ts",
+    "build-ui": "npm run build --workspace ui/ui-icons --workspace ui/ui-core --workspace ui/ui-charts --workspace ui/ui-warped-map"
```
```suggestion
      "build-ui": "npm run build --workspaces --if-present --exclude ui/base --exclude ui/storybook"
```

emersion replied:
`npm run` doesn't have an `--exclude` flag.

Yohh commented on front/ui/.gitignore:
```diff
@@ -0,0 +1,13 @@
+*.tgz
+storybook-static
+
+# Direnv
+.envrc
+.direnv
```
already in root gitignore, don't think it,s needed here

emersion replied:
Indeed, dropped!

Yohh commented on front/ui/README.md:
```diff
@@ -0,0 +1,66 @@
+
+# Build Instructions
+
+1. Run `npm install`
+2. Run `npm run build`
```
I didn't need it after running `/front` `npm run build-ui`, I also think this section needs some extra comments to explain how it works now it's integrated in osrd main repo

emersion replied:
Indeed, I changed it to `npm run build-ui` here to only build osrd-ui instead of osrd + storybook.

What exactly would you want to add here? I think with these updated commands it should work the same as before?

Yohh replied:
I just wonder if it's a good thing to keep those instructions here, this script is from upper package.json, in `front/` which already has a ReadMe with setup instructions...

emersion replied:
I think it's still useful to have instructions to work on osrd-ui alone, without touching/building osrd itself, especially for external contributors. This README is a good place for that IMHO, to avoid having these users have to go through a lot of other osrd-specific instructions and figure out which they should ignore.

Yohh replied:
ok fine

Yohh commented on front/vite.config.ts:
```diff
@@ -25,8 +25,7 @@ export default defineConfig(({ mode }) => {
         eslint: {
-          lintCommand:
-            'eslint --cache --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --max-warnings 0',
+          lintCommand: 'eslint --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --max-warnings 0',
```
why did we add it?

emersion replied:
See the commit description for "front: disable ESLint cache". I added `--cache` a while back to improve ESLint's startup time.

Yohh commented:
> To fix, run `npm run build --workspaces` before starting the container, or restart the container (either should work).

what about updating `osrd-compose` script for `dev-front` mode and run this command?

emersion commented:
> what about updating osrd-compose script for dev-front mode and run this command?

This would delay the container startup by 30s or 1min every time it's started, which was deemed more annoying.

emersion committed:
front: trim down osrd-ui's .gitignore

The direnv and IDE stuff is already covered by the root .gitignore
file.

Signed-off-by: Simon Ser <contact@emersion.fr>

clarani commented on front/ui/README.md:
```diff
@@ -10,8 +10,7 @@ Development will leverage preconstruct. **Documentation yet to come**.
 
-In a nutshell: we consider our version numbers metadata of the project. We use
-[Semantic Versioning](https://semver.org/).
+osrd-ui releases are published automatically together with OSRD releases. osrd-ui's latest version is always in sync with OSRD.
```
This might affect our external users :thinking: But I remember the discussions about it :+1:

emersion replied:
Yeah. We can change how releases are managed later on if needed.

clarani commented on front/docker/Dockerfile.devel:
```diff
@@ -23,4 +23,4 @@ ENV VITE_OSRD_GIT_DESCRIBE=${OSRD_GIT_DESCRIBE}
 # Start the app
 COPY docker/dev-entrypoint.sh /
 ENTRYPOINT ["/dev-entrypoint.sh"]
-CMD ["sh", "-c", "id && npm install && exec npm run start-container"]
```
Only small questions by curiosity :) 
Why do you remove the `exec` ? Adding the `()` run the 2 commands in parallel ?

emersion replied:
`exec` is a shell built-in utility which replaces the shell with another command: the shell's process is re-used for executing the supplied command. With `npm run start`, there are two processes, the shell and npm as a child process. With `exec npm run start`, there is only a single process, npm.

`&` runs the previous command in parallel with the rest. When evaluating `a & b`, the shell forks, evaluates `a` in the child, then evaluates `b` in the parent.

`exec` can only be used to execute a single process. To execute two commands in parallel, the shell needs to remain (not be replaced by another command) to coordinate everything.

The `()` is to disambiguate what runs in parallel. With `a && b & c` it's unclear whether `a && b` runs in parallel with `c`, or if `a` is executed and then `b` and `c` run in parallel. (Only folks who wrote a POSIX shell implementation remember that without the parentheses, `a && b` runs in parallel with `c`.)

clarani replied:
Ok thanks :)

clarani commented on front/package.json:
```diff
@@ -3,15 +3,23 @@
   "private": true,
   "type": "module",
+  "workspaces": [
+    "ui/base",
```
What `ui/base` is used for ?

emersion replied:
See the section "Why keep front/ui/base/package.json?" in the PR description :P

theocrsb commented on front/ui/README.md:
```diff
@@ -0,0 +1,77 @@
+package.json will also prevent the package from being published.
+
+# Build Instructions
+
```
I don't know if this is very useful, but perhaps we could specify the folder in which to run these commands, as in the `README` of `OSRD`

emersion replied:
Sure, I've added an extra bullet point to make this clearer!

theocrsb commented:
We have an error + white screen on `Docs` pages

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/fcb77ce4-4e69-4295-b0c6-19a95ecbd1a5" />

<img width="312" alt="image" src="https://github.com/user-attachments/assets/c45065a5-c89e-4172-a9d2-a48f30e53b67" />

theocrsb commented on front/ui/README.md:
```diff
@@ -0,0 +1,66 @@
+2. Run `npm run build-ui`
+3. Run `cd storybook` and `npm start`
+
+While developing on a project, you need to run the following commands to see the css modifications:
+
+1. Run `cd ui-<name-of-the-project>`
+2. Run `npm run watch`
```
Personally, I didn't need to run the `watch` command to see the changes :)

emersion replied:
I don't understand how that could be. Without `watch`, the files in `dist/` are never updated.

theocrsb replied:
Yet we did the test with @clarani. but only in storybook

emersion replied:
What I did:

- `cd front/ui/storybook` then `npm run start`
- Open the manchette story in the browser
- Edit `ui/ui-charts/src/manchette/components/Waypoint.tsx` to add "SALUT THÉO" in `.waypoint-name`
- The story isn't updated
- In another terminal, `cd front/ui/ui-charts` then `npm run watch`
- The story is updated

theocrsb replied:
indeed, thank you!

kmer2016 reviewed (commented):
I started the stack using Docker:
```
./osrd-compose dev-front build    
./osrd-compose up -d
```
Hot reload works for changes in the osrd codebase, but it doesn't for changes in osrd-ui.

emersion commented:
> Hot reload works for changes in the osrd codebase, but it doesn't for changes in osrd-ui.

This is expected. Hot reload for osrd-ui requires more changes and is trickier to implement (because of the relative dependencies between osrd-ui packages).

Hot reload should work if you start `npm run watch` in the osrd-ui package you're working on.

clarani commented on front/package.json:
```diff
@@ -145,13 +153,18 @@
-    "build": "vite build",
-    "test": "vitest",
-    "lint": "eslint --cache --ext .ts,.tsx,.js,.jsx src scripts tests *.ts --max-warnings 0",
+    "build": "npm run build --workspaces --if-present && vite build",
```
Do we need to build storybook here too ?

emersion replied:
Ah, good point, I didn't explain the motivation. I've added a new section "Should we build osrd-ui's storybook when building osrd's frontend?" in the PR description. Happy to change it if you prefer.

emersion commented:
> We have an error + white screen on Docs pages 

I could not reproduce this bug. What I did, from a clean slate:

```
cd front
npm install
npm run build-ui
cd ui/storybook
npm run start
```

flomonster reviewed (approved):
LGTM.

Note: I found some refs to "https://github.com/osrd-project/osrd-ui/". Do you think we should replace it in this PR?

We will need some follow-up PR to clean up CI files.

emersion committed:
front: adjust osrd-ui URLs in package.json files

Switch these to the osrd repository.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
> I found some refs to "https://github.com/osrd-project/osrd-ui/". Do you think we should replace it in this PR?

Good catch, I've added a commit to fixup these.

SharglutDev commented on front/ui/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## Publishing versions
+
+osrd-ui releases are published automatically together with OSRD releases. osrd-ui's latest version is always in sync with OSRD.
```
This has probably been answered already but what if, when publishing a new osrd release, the last osrd-ui release hasn't been bumped on osrd ? (or, if we don't release osrd-ui individually anymore, the last osrd-ui changes haven't been added to osrd)

emersion replied:
OSRD always uses the local osrd-ui directory instead of the NPM package. If a breaking change is made to osrd-ui, OSRD needs to be updated to account for the change or else the build fails.

ghost commented:
Thank you very much, I mostly checked the GitHub CI part, but overall it looks good to me.

SharglutDev reviewed (approved):
Thank you for the great work ! (not tested)

kmer2016 reviewed (approved):
Very good ! Thank you for the PR !

emersion committed:
github: temporarily disable automatic osrd-ui releases

We need a bit more time to come up with a plan.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
Rebased and updated for osrd-ui v0.0.73.

I've also appended a commit to temporarily disable automatic releases for now:
https://github.com/OpenRailAssociation/osrd/pull/10711/commits/4ab6c34b362561048fffe6a14bfb2ee23b8e4085

clarani reviewed (approved):
LGTM, tested :white_check_mark:

emersion commented:
Trivial rebase to bump osrd-ui to v0.0.74.
